xmtp_mls/src/groups/device_sync/backup/backup_importer.rs

```diff
@@ -0,0 +1,135 @@
+use prost::Message;
+use sha2::digest::{generic_array::GenericArray, typenum};
+use std::pin::Pin;
+use xmtp_proto::xmtp::device_sync::{backup_element::Element, BackupElement, BackupMetadata};
+
+use super::BackupError;
+
+#[cfg(not(target_arch = "wasm32"))]
+mod file_import;
+
+pub struct BackupImporter {
+    pub metadata: BackupMetadata,
+    decoded: Vec<u8>,
+    decoder: ZstdDecoder<Pin<Box<dyn AsyncBufRead + Send>>>,
+
+    cipher: AesGcm<Aes256, typenum::U12, typenum::U16>,
+    nonce: GenericArray<u8, typenum::U12>,
+}
+
+impl BackupImporter {
+    pub(super) async fn load(
+        mut reader: Pin<Box<dyn AsyncBufRead + Send>>,
+        key: &[u8],
+    ) -> Result<Self, DeviceSyncError> {
+        let mut nonce = [0; NONCE_SIZE];
+        reader.read_exact(&mut nonce).await?;
+
+        let mut importer = Self {
+            decoder: ZstdDecoder::new(reader),
+            decoded: vec![],
+            metadata: BackupMetadata::default(),
+
+            cipher: Aes256Gcm::new(GenericArray::from_slice(key)),
+            nonce: GenericArray::from(nonce),
+        };
+
+        let Some(BackupElement {
+            element: Some(Element::Metadata(metadata)),
+        }) = importer.next_element().await?
+        else {
+            return Err(BackupError::MissingMetadata)?;
+        };
+
+        importer.metadata = metadata;
+        Ok(importer)
+    }
+
+    async fn next_element(&mut self) -> Result<Option<BackupElement>, DeviceSyncError> {
+        let mut buffer = [0u8; 1024];
+        let mut element_len = 0;
+        loop {
+            let amount = self.decoder.read(&mut buffer).await?;
+            self.decoded.extend_from_slice(&buffer[..amount]);
+
+            if element_len == 0 && self.decoded.len() >= 4 {
+                let bytes = self.decoded.drain(..4).collect::<Vec<_>>();
+                element_len = u32::from_le_bytes(bytes.try_into().expect("is 4 bytes")) as usize;
+            }
+
+            if element_len != 0 && self.decoded.len() >= element_len {
+                let decrypted = self
+                    .cipher
+                    .decrypt(&self.nonce, &self.decoded[..element_len])?;
+                let element = BackupElement::decode(&*decrypted);
+                self.decoded.drain(..element_len);
+                return Ok(Some(
+                    element.map_err(|e| StorageError::Deserialization(e.to_string()))?,
+                ));
+            }
+
+            if amount == 0 && self.decoded.is_empty() {
+                break;
+            }
+        }
+
+        Ok(None)
+    }
+
+    pub async fn insert(&mut self, provider: &XmtpOpenMlsProvider) -> Result<(), DeviceSyncError> {
+        provider
+            .transaction_async(|provider| async move {
+                let conn = provider.conn_ref();
+                while let Some(element) = self
+                    .next_element()
+                    .await
+                    .map_err(|e| StorageError::Generic(e.to_string()))?
```

Comment:
I'd rather stay away from Generic string errors if possible, they can't be matched on and more difficult to process afterwards than normal run of the mill errors. Both these functions return DeviceSyncError, why not use that?

Reply:
I would love for this to return a DeviceSync error, but since it's in a transaction closure, it enforces StorageError. I'll see if I can put some more through towards it to get out of this pattern.

Reply:
Oh, I completely overlooked the fact that `async_transaction` returns an unconstrained generic. It's fixed.